addon-resizer/nanny/kubernetes_client.go

```diff
@@ -20,40 +20,59 @@ import (
 	"fmt"
 	"time"
 
-	api "k8s.io/kubernetes/pkg/api"
-	apiv1 "k8s.io/kubernetes/pkg/api/v1"
-	cache "k8s.io/kubernetes/pkg/client/cache"
-	client "k8s.io/kubernetes/pkg/client/clientset_generated/release_1_3"
-	runtime "k8s.io/kubernetes/pkg/runtime"
-	wait "k8s.io/kubernetes/pkg/util/wait"
-	watch "k8s.io/kubernetes/pkg/watch"
+	apps "k8s.io/api/apps/v1"
+	api "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/labels"
+	kube_client "k8s.io/client-go/kubernetes"
+	kube_client_apps "k8s.io/client-go/kubernetes/typed/apps/v1"
+	v1appslister "k8s.io/client-go/listers/apps/v1"
+	v1lister "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
 )
 
 type kubernetesClient struct {
-	namespace  string
-	deployment string
-	pod        string
-	container  string
-	clientset  *client.Clientset
-	nodeStore  cache.Store
-	reflector  *cache.Reflector
+	nodeLister       v1lister.NodeLister
+	podLister        v1lister.PodNamespaceLister
+	deploymentLister v1appslister.DeploymentNamespaceLister
+	deploymentClient kube_client_apps.DeploymentInterface
+	namespace        string
+	deployment       string
+	pod              string
+	container        string
+	stopChannel      chan struct{}
 }
 
-func (k *kubernetesClient) CountNodes() (uint64, error) {
-	err := wait.PollImmediate(time.Second, time.Minute, func() (bool, error) {
-		if k.reflector.LastSyncResourceVersion() == "" {
-			return false, nil
-		}
-		return true, nil
-	})
-	if err != nil {
-		return 0, err
+// NewKubernetesClient gives a KubernetesClient with the given dependencies.
+func NewKubernetesClient(kubeClient kube_client.Interface, namespace, deployment, pod, container string) KubernetesClient {
+	stopChannel := make(chan struct{})
+	result := &kubernetesClient{
+		stopChannel:      stopChannel,
+		namespace:        namespace,
+		deployment:       deployment,
+		pod:              pod,
+		container:        container,
+		nodeLister:       newReadyNodeLister(kubeClient, stopChannel),
+		podLister:        newPodListerByNamespace(kubeClient, stopChannel, namespace),
+		deploymentLister: newDeploymentListerByNamespace(kubeClient, stopChannel, namespace),
+		deploymentClient: kubeClient.AppsV1().Deployments(namespace),
 	}
-	return uint64(len(k.nodeStore.List())), nil
+	return result
 }
 
-func (k *kubernetesClient) ContainerResources() (*apiv1.ResourceRequirements, error) {
-	pod, err := k.clientset.CoreClient.Pods(k.namespace).Get(k.pod)
+func (k *kubernetesClient) Stop() {
+	for i := 0; i < 3; i++ {
```

Comment:
The i < 3 seems really error prone - if someone adds another lister, no way they will know thay have to increase it. Can you add a stopChan per lister instead?